### ISSUE TITLE
”Show password” button should not be active on passwords autofilled by the browser

### ISSUE THREAD

**Issue by simevidas** [alphagov/govuk_publishing_components#2182]
GOV.UK blog recently published the article ”[Simple things are complicated: making a show password option](https://technology.blog.gov.uk/2021/04/19/simple-things-are-complicated-making-a-show-password-option/)” which discusses the new “Show password” functionality on the login page of GOV.UK Accounts.

It is my opinion that this functionality should *not* be active if the password field is autofilled by the browser because  it makes it much easier for another person to *quickly* see your password.

1. Go to website
2. Log out
3. Go to login page
4. Click ”Show password”

This only takes 10 seconds, and anyone can do it. Here is a video demonstration:

https://user-images.githubusercontent.com/716405/124346203-ddc71800-dbdd-11eb-9d2d-a0c5c5ab6ded.mp4

(No worries, I have changed my password before uploading this video.)

You could argue that if somebody else gains temporary access to your computer and browser, they could extract your GOV.UK password from the browser itself, but this is not necessarily the case. For example, Firefox now protects the user’s login information via biometric authentication:

<img width="891" alt="Screen Shot 2021-07-03 at 8 59 54 AM" src="https://user-images.githubusercontent.com/716405/124346156-8cb72400-dbdd-11eb-85fc-af375981e75d.png">

To sum up, I don’t think “Show password” on autofilled passwords is a good idea. It should be disabled if possible.

**Comment by andysellick** [comment]
Thanks for raising this issue. Unfortunately this is an inherent flaw with browser password managers. If an attacker has gained physical access to your system, any stored password can be accessed through the browser with this method. This can even be achieved with one more step regardless of whether the password field has a show password option or not (I won't specifically explain this step).

We try to make our systems as secure as possible. In this case we can't do anything if someone's system has been compromised to this level - it's not the fault of the show password option.

**Comment by simevidas** [comment]
@andysellick As the second screenshot in my post shows, Firefox’s password manager asks for biometric authentication to reveal website passwords. So even if an attacker gains direct access to the browser, they won’t be able to extract passwords from Firefox’s password manager.

This leaves the website itself, as you said. If the browser autofills the password on the website’s login page, then the attacker can view the password by changing the password field’s `<input type>` attribute.

So while revealing autofilled passwords is possible, your “Show” button makes it *much quicker easier*. That is my argument. The person does not need to know HTML and DevTools to be able to view the victim’s password quickly by logging out and going to the login page.

Why make this so easy? Is there a reason why “Show” needs to be active for autofilled passwords? The purpose of ”Show” is to help with *entering* passwords. If the user wants to view their password, they should go into their password manager where the password is secure.

**Comment by taylor-hunt-kr** [comment]
I tried implementing the sort of safeguard @simevidas is describing on a website before, and unfortunately [detecting autofill in browsers is such an unconscionable, unreliable hassle](https://gist.github.com/jonathantneal/d462fc2bf761a10c9fca60eb634f6977) that we had to back it out altogether. (In particular, High-Contrast Mode, which GOV.UK wants to do right by.)

**Comment by simevidas** [comment]
@taylor-hunt-kr I might be missing something, but the `:autofill` pseudo-class seems to work well for this. 

```css
:autofill + button { 
  display: none;
}
```

Result in Firefox:


https://user-images.githubusercontent.com/716405/125005582-67eb0280-e05c-11eb-8dbe-64e71b5f042e.mp4

**Comment by taylor-hunt-kr** [comment]
That pseudo-class is new to me! Doesn’t fit our browser support, sadly, but at least it gets us another 90% there.

**Comment by andysellick** [comment]
The pseudo class is a good idea, but it isn't supported by every browser that we support, and any smart attacker can get around it anyway using the techniques already discussed.

We'll continue to iterate and will keep an eye out for a possible solution in future, but for now I'm closing this issue. Thanks again for raising.
